### PR TITLE
Bug fix: Defining and using Jinja macro in the same file causes runtime error

### LINKED ISSUE #2456: jinja templater: Defining and using a macro in the same file causes runtime error
## Expected Behaviour
Successful linting
## Observed Behaviour
Runtime error:
```
(sqlfluff-3.10.1) ➜  sqlfluff git:(main) ✗ sqlfluff lint test.sql
WARNING    Unable to lint test.sql due to an internal error. Please report this as an issue with your query's contents and stacktrace below!
To hide this warning, add the failing file to .sqlfluffignore
Traceback (most recent call last):
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/linter/runner.py", line 103, in run
    yield partial()
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/linter/linter.py", line 598, in lint_rendered
    parsed = cls.parse_rendered(rendered)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/linter/linter.py", line 352, in parse_rendered
    tokens, lvs, config = cls._lex_templated_file(
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/linter/linter.py", line 139, in _lex_templated_file
    tokens, lex_vs = lexer.lex(templated_file)
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/parser/lexer.py", line 321, in lex
    segments: Tuple[RawSegment, ...] = self.elements_to_segments(
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/parser/lexer.py", line 348, in elements_to_segments
    source_slice = templated_file.templated_slice_to_source_slice(
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/templaters/base.py", line 294, in templated_slice_to_source_slice
    ts_stop_sf_start, ts_stop_sf_stop = self._find_slice_indices_of_templated_pos(
  File "/Users/bhart/dev/sqlfluff/src/sqlfluff/core/templaters/base.py", line 180, in _find_slice_indices_of_templated_pos
    raise ValueError("Position Not Found")
ValueError: Position Not Found
```
## Steps to Reproduce
Lint this file:
```
{% macro get_keyed_nulls(columns) %}

  COALESCE({{columns}}, MD5(-1))

{% endmacro %}

SELECT
    c1,
    {{ get_keyed_nulls("other_id") }} AS other_id
FROM my_table
```
## Configuration
```
Include your SQLFluff configuration here
```

### PR DESCRIPTION
**Pull request by barrywhart**
### Brief summary of the change made
Fixes #2456

### Pull Request checklist
- [x] Please confirm you have completed any of the necessary steps below.

- Included test cases to demonstrate any code changes, which may be one or more of the following:
  - `.yml` rule test cases in `test/fixtures/rules/std_rule_cases`.
  - `.sql`/`.yml` parser test cases in `test/fixtures/dialects` (note YML files can be auto generated with `tox -e generate-fixture-yml`).
  - Full autofix test cases in `test/fixtures/linter/autofix`.
  - Other.
- Added appropriate documentation for the change.
- Created GitHub issues for any relevant followup/future enhancements if appropriate.

### CODE REVIEW

File: src/sqlfluff/core/templaters/slicers/tracer.py [hunk 1]
```diff
@@ -319,7 +326,10 @@ def _slice_template(self) -> List[RawFileSlice]:
                                 f"\0{prefix}{unique_alternate_id} {open_} "
                                 f"{trimmed_content} {close_}"
                             )
-                if block_type == "block_start" and trimmed_content.split()[0] == "set":
+                if block_type == "block_start" and trimmed_content.split()[0] in (
+                    "macro",
+                    "set",
+                ):
```
Comment (barrywhart) [inline src/sqlfluff/core/templaters/slicers/tracer.py:332]: To fix this bug, we need to treat macro definitions the same as `{% set %}` blocks.

File: src/sqlfluff/core/templaters/slicers/tracer.py [hunk 2]
```diff
@@ -99,12 +99,18 @@ def trace(self) -> JinjaTrace:
...
-                self.record_trace(slice_length, target_slice_idx)
+                # If we find output from a {% set %} directive or a macro,
+                # record a trace without reading or updating the program
+                # counter. Such slices are always treated as "templated"
+                # because they are inserted during expansion of templated
+                # code (i.e. {% set %} variable or macro defined within the
+                # file).
+                self.record_trace(
+                    slice_length, target_slice_idx, slice_type="templated"
+                )
```
Comment (barrywhart) [inline src/sqlfluff/core/templaters/slicers/tracer.py:113]: This fixes a different but related bug: Code inserted by using a `{% set %}` variable **or** a macro defined in the same file should be labelled as "templated". I discovered this when adding the L009 test case -- initially it was seeing literal newlines inserted by a macro as actual literals (bad!).

File: src/sqlfluff/core/templaters/slicers/tracer.py [hunk 3]
```diff
@@ -154,11 +160,12 @@ def move_to_slice(self, target_slice_idx, target_slice_length):
                 candidates.sort(key=lambda c: abs(target_slice_idx - c))
                 self.program_counter = candidates[0]
 
-    def record_trace(self, target_slice_length, slice_idx=None):
+    def record_trace(self, target_slice_length, slice_idx=None, slice_type=None):
```
Comment (barrywhart) [inline src/sqlfluff/core/templaters/slicers/tracer.py:163]: Adds a new, optional parameter that lets callers override the slice type for a trace entry.

**Review by tunetheweb — APPROVED** [review]
LGTM!

**Pull request merged**